draft-yasskin-dispatch-web-packaging.md

```diff
@@ -0,0 +1,650 @@
+
+Load one CBOR item starting at *manifest-start* as a `signed-manifest` from the
+following CDDL:
+
+~~~~~ cddl
+signed-manifest = {
+  manifest: manifest,
+  certificates: [+ certificate],
+  signatures: [+ signature]
+}
+
+manifest = {
+  metadata: manifest-metadata,
+  resource-hashes: {* hash-algorithm => [hash-value]},
+  ? subpackages: [* subpackage],
+}
+
+manifest-metadata = {
+  date: time,
+  origin: uri,
+  * tstr => any,
+}
+
+; From https://www.w3.org/TR/CSP3/#grammardef-hash-algorithm.
+hash-algorithm /= "sha256" / "sha384" / "sha512"
+; Note that a hash value is not base64-encoded, unlike in CSP.
+hash-value = bstr
+
+; X.509 format; see https://tools.ietf.org/html/rfc5280
+certificate = bstr
+
+signature = {
+  ; This is the index of the certificate within the
+  ; certificates array to use to validate the signature.
+  keyIndex: uint,
+  signature: bstr,
+}
+~~~~~
+
+If the item doesn't match the CDDL or it's not a Canonical CBOR item (Section
+3.9 of {{CBOR}}), parsing MUST fail.
+
+Parse the elements of `certificates` as X.509 certificates within the
+{{!RFC5280}} profile. If any certificate fails to parse, parsing MUST fail.
+
+Let *message* be the concatenation of the following byte strings. This matches
+the {{?TLS1.3=I-D.ietf-tls-tls13}} format to avoid cross-protocol attacks when
+TLS certificates are used to sign manifests.
+
+1. A string that consists of octet 32 (0x20) repeated 64 times.
+1. A context string: the ASCII encoding of "Web Package Manifest".
+1. A single 0 byte which serves as a separator.
+1. The bytes of the `manifest` CBOR item.
+
+Let *signing-certificates* be an empty array.
+
+For each element *signature* of `signatures`:
+
+1. Let *certificate* be `certificates`\[*signature*\["keyIndex"]].
+
+1. The parser MUST define a partial function from public key types to signing
+   algorithms, with the following map as a subset:
+
+   RSA, 2048 bits:
+   : rsa_pss_sha256 as defined in Section 4.2.3 of {{!TLS1.3}}
+
+   EC, with the secp256r1 curve:
+   : ecdsa_secp256r1_sha256 as defined in Section 4.2.3 of {{!TLS1.3}}
+
+   EC, with the secp384r1 curve:
+   : ecdsa_secp384r1_sha384 as defined in Section 4.2.3 of {{!TLS1.3}}
+
+   Let *signing-alg* be the result of applying this function to the key type in
+   *certificate*'s Subject Public Key Info. If the function is undefined on this
+   input, the parser MUST continue to the next *signature*.
+
+1. Use *signing-alg* to verify that *signature*\["signature"] is *message*'s
+   signature by *certificate*'s public key. If it's not, the parser MUST
+   continue to the next *signature*.
+
+1. Append *certificate* to *signing-certificates*. Note that failed signatures
+   simply cause their certificate to be ignored, so that packagers can give new
+   signature types to parsers that understand them.
+
+
+Let *origin* be `manifest`\["metadata"]\["origin"].
+
+Try to find a certificate in *signing-certificates* that has an identity
```

Comment:
"Try to find" is unclear, perhaps something like "iterate through signing-certificates until one is found that has an identity matching ...